Add error preview tooltip to error log row, make exception message property multiline

adamint opened the pull request:
Fixes #1177 

Adds a `FluentTooltip` that contains 1) an error preview that includes the exception type and message, and 2) a button through which you can copy that preview. Also adds a property to `PropertyGrid` that allows specific properties to be marked as wrappable (`WrappedProperties`), and adds text wrapping to `ex.Message`

![image](https://github.com/dotnet/aspire/assets/20359921/1545e1d1-d6d9-4f2a-92e0-3eeb143dff2b)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2087)

Commit:
add tooltip with error info to log

Commit:
wip

Commit:
Merge branch 'main' into dev/adamint/1177-better-inline-exceptions

Commit:
make ex.Message multiline grid value, implement copy in error tooltip

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor:
```diff
@@ -0,0 +1,52 @@
+
+@if (TryGetErrorInformation(out var errorInfo))
+{
+    var test = Guid.NewGuid().ToString();
```
Better name

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:
```diff
@@ -0,0 +1,34 @@
+
+public partial class LogLevelColumnDisplay
+{
+    private bool TryGetErrorInformation([MaybeNullWhen(false)] out string errorInfo)
```
```suggestion
    private bool TryGetErrorInformation([NotNullWhen(true)] out string? errorInfo)
```
Isn't this the idomatic way to do this? I've never seen the opposite.

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor:
```diff
@@ -0,0 +1,52 @@
+    <FluentIcon Id="@id" Icon="Icons.Filled.Size16.Warning" Color="Color.Warning" Class="severity-icon"/>
+}
+
+@if (TryGetErrorInformation(out var errorInfo))
```
This came up before: exceptions aren't tied to the severity. It's possible to have an exception at any level. The icon shouldn't be used for this.

Also, how would anyone know to hover the cursor over the icon to get exception detail?

JamesNK replied:
What about adding an icon to indicate the log entry has exception data at the end of the message? Where the red box is:

![image](https://github.com/dotnet/aspire/assets/303201/e800724b-2033-4044-a32b-f6676529bf54)

Hovering over that shows the error detail popup.

Also, while we're here, there should be a copy button for the message, like there is for the Source column on the resources page.

JamesNK commented on src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor:
```diff
@@ -18,6 +19,7 @@
                       ValueSort="_valueSort"
-                      HighlightText="@_filter" />
+                      HighlightText="@_filter"
+                      WrappedProperties="@ImmutableArray.Create("ex.Message")"/>
```
What is the special meaning of `ex.Message`?

These are the standard OTEL properties:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/ebb35248003885c6a3315417d090fb081fccd2b0/src/Shared/SemanticConventions.cs#L96-L100

Can you investigate why you have an exception with non-standard properties.

JamesNK replied:
I don't think hardcoding known properties is a good long term situation here. There are many tools that display structured logs out there. What do they do in their UI?

I feel like there should be a way to select a property row and have it switch to multiline, but I don't have another tool that does that off the top of my head as prior art.

adamint replied:
> I feel like there should be a way to select a property row and have it switch to multiline text box that preserves whitespace

I have not checked others, but this does not exist in the msbuild log viewer; it displays as one line.

The solution for now should then just be to wrap all properties. No need to add a toggle

JamesNK replied:
msbuild log viewer is probably made for devs, by devs. We should look at what other tools in this space do. What does AppInsights in azure do? Or datadog, seq, etc.

davidfowl replied:
There are others that we will need to be expandable. Can it be based on the content truncation?

adamint replied:
> Can it be based on the content truncation

possibly, but I don't know how to do this

JamesNK commented on src/Aspire.Dashboard/wwwroot/js/app.js:
```diff
@@ -70,7 +70,7 @@ window.copyTextToClipboard = function (id, text, precopy, postcopy) {
             checkmarkIcon.style.display = 'inline';
         })
-        .catch(() => {
+        .catch((e) => {
```
What is this for? You could write it to the console.log if it's not already written.

JamesNK commented on src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor.css:
```diff
@@ -54,3 +54,7 @@
 }
+
+::deep .wrap {
+    text-wrap: wrap;
```
This wraps content. But it doesn't respect any of the whitespace (line breaks, indentation) in the content.

For example, this exception message is just one long wrapped line without applying any of the content's whitespace:

![image](https://github.com/dotnet/aspire/assets/303201/2b478ffc-a49f-4f0b-a228-4e541bcf5774)

adamint replied:
responded [here](https://github.com/dotnet/aspire/pull/2087#discussion_r1480422514)

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    private bool TryGetErrorInformation([MaybeNullWhen(false)] out string errorInfo)
+    {
+        if (LogEntry.Properties.GetValue("ex.Type") is { } exceptionType)
```
Non-standard property names. Could fallback to them, e.g. look for exception.type first and then ex.Type. Need to figure out why ex.Type is here at all first.

adamint replied:
@Erarndt is there a reason these property names are used? They come from here: https://github.com/dotnet/aspire/blob/0282715bfee52b3ed982228a67811b62ec8cb6b3/src/Components/Aspire.RabbitMQ.Client/RabbitMQEventSourceLogForwarder.cs#L156

adamint replied:
Falling back to them

JamesNK replied:
Create an issue for this.

JamesNK commented:
Popup displays off the edge of the page:

![image](https://github.com/dotnet/aspire/assets/303201/391734f7-138e-4a53-a06a-c0b939f2aa4e)

Maximum height is needed.

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:
```diff
@@ -0,0 +1,34 @@
+        {
+            var message = LogEntry.Properties.GetValue("ex.Message");
+            var stackTrace = LogEntry.Properties.GetValue("ex.StackTrace");
+            errorInfo = LogEntry.Message + Environment.NewLine + $"{exceptionType}: {message}" + Environment.NewLine + stackTrace;
```
```suggestion
            errorInfo = $"{exceptionType}: {message}" + Environment.NewLine + stackTrace;
```
Just include the exception information. There should be a separate way to copy the log message.

JamesNK replied:
Actually, from what I see of the data and according to the [spec](https://github.com/open-telemetry/semantic-conventions/blob/45a03e67ce2e109e860c38c60edf6367a1124545/docs/attributes-registry/exception.md), `exception.stacktrace` already includes the exception type and message.

errorInfo should be _just_ `exception.stacktrace`/`ex.StackTrace`.

JamesNK commented:
I suggest this PR just make changes to the log entry grid. Leave changes to the details view to another PR.

adamint commented:
They're both related changes, as part of the same experience; IMO should stay in the same PR

adamint commented:
Made the popup slightly wider, added max height with scrolling.

![image](https://github.com/dotnet/aspire/assets/20359921/9e89a65f-1744-4f2e-982c-3617b647f389)

Preserves whitespace in property grid
![image](https://github.com/dotnet/aspire/assets/20359921/f727f982-4e17-4563-9ff9-48bc70fc4c78)

Commit:
move popup to icon in message

Commit:
remove vertical-align center from log message

JamesNK commented:
> Made the popup slightly wider, added max height with scrolling. I kept the icon on the left of the column, as it conforms with other columns.
> 
> ![image](https://private-user-images.githubusercontent.com/20359921/302759632-9e89a65f-1744-4f2e-982c-3617b647f389.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNjYxOTMsIm5iZiI6MTcwNzI2NTg5MywicGF0aCI6Ii8yMDM1OTkyMS8zMDI3NTk2MzItOWU4OWE2NWYtMTc0NC00ZjJlLTk4MmMtMzYxN2I2NDdmMzg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA3VDAwMzEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNjFiMDkwNTg5OTBmOTRmMDJmMTk4MDc4NTI4YTMzY2VjYWViMDJiMWIxMDI4OGYwYjI0MDIwMDBkZWViZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KTv08HPHAxRVFU5aQmyz5iBOHu78cgueQIAGkuB2Qeo)

Ok. I'd prefer it if the icon color was white (or whatever the font color is). A logged exception isn't necessarily a bad thing.

I think the tooltip needs a title, e.g. "<h3>Exception details</h3>"

> 
> Preserves whitespace in property grid ![image](https://private-user-images.githubusercontent.com/20359921/302759696-f727f982-4e17-4563-9ff9-48bc70fc4c78.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyNjYxOTMsIm5iZiI6MTcwNzI2NTg5MywicGF0aCI6Ii8yMDM1OTkyMS8zMDI3NTk2OTYtZjcyN2Y5ODItNGUxNy00NTYzLTlmZjktNDhiYzcwZmM0Yzc4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA3VDAwMzEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjYmNhY2YyYjgyNjdkYWQwYmUwOWMwZTZlODdiZGFlMjBhYWNiNGE0Mjg3ZGZkYTJiODNjOWJhN2RjNTYyZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BgU17-CPw4Z_4uo9hRLuaZ_nII8LkP0C2tN1XhrqhD4)

It looks ok. When preserving whitespace, you need to be careful about what whitespace gets accidentally added in the HTML from identing tags.

Remove this whitespace here:
![image](https://github.com/dotnet/aspire/assets/303201/dac84a2b-11d6-4155-91be-cd5c54474584)

JamesNK commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            var message = GetProperty("exception.message") ?? GetProperty("ex.Message");
+            var type = GetProperty("exception.type") ?? GetProperty("ex.Type");
+            errorInfo = LogEntry.Message + Environment.NewLine + $"{type}: {message}" + Environment.NewLine + stackTrace;
```
```suggestion
            errorInfo = $"{type}: {message}" + Environment.NewLine + stackTrace;
```
Exception details doesn't include log enetry message.

Commit:
Merge branch 'main' into dev/adamint/1177-better-inline-exceptions

adamint commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor:
```diff
@@ -0,0 +1,51 @@
+
+    <FluentTooltip Anchor="@iconId" MaxWidth="650px">
+        <div style="padding: 10px">
+            <pre style="white-space: pre-wrap; max-height: 400px; overflow: auto;">@errorInfo</pre>
```
changed

Commit:
add title, remove extra spacing in property grid

Commit:
change error info icon to accent color

adamint commented:
![image](https://github.com/dotnet/aspire/assets/20359921/e11c1e2c-171b-491d-a021-b62c793b4a12)
![image](https://github.com/dotnet/aspire/assets/20359921/b6a6721e-658a-4898-a530-8037e2645f9c)
Added title, removed the extra spacing that you pointed out James. I don't think the icon should be the same color as the text, as it will not stand out, but how is accent?

JamesNK commented:
> but how is accent?

Yeah that's good.

JamesNK reviewed (approved):
👍 

An odd thing I see is the scroll position is maintained between the tooltip closing and reopened. It feels like a bug to me but maybe some people would want that? IDK

adamint commented:
> 👍
> 
> An odd thing I see is the scroll position is maintained between the tooltip closing and reopened. It feels like a bug to me but maybe some people would want that? IDK

That seems like a positive to me, but I don't know. Happy to change if we get negative feedback on it

dersia commented:
thanks for the copy to clipboard button!
